Refactored AM modules for Basic Sprint 3 Prototype Requirements and Added Tests for Get_Supported_Actions Module

MTUC08 opened the pull request:
For Sprint 3, we had to refactor our modules to meet the basic sprint 3 prototype requirements ( #115 ) and add unit tests for all of our functions. 

This PR includes the refactored code and tests for the get_actions module (which is used by other teams #150 ), and will be followed by a PR including the rest of our tests, so as to attempt to minimize the size of the PR to master. Due to dependency on game-states functions, we could not begin this PR until this afternoon. The other tests (#169 #180 #181 #182 #241 #242 ) will be added to master after this merge.

MTUC08 commented:
Since this PR will directly affect CLI, who will use our altered functions, I'm planning on adding a member of that team as a reviewer instead.

ecoble commented on src/action_management/Makefile:
```diff
@@ -3,25 +3,29 @@
+CFLAGS = -Wall -Wextra -O2 -g -I./include/ -I../common/include/ -I../game-state/include/
 OBJS = $(SRCS:.c=.o)
-LIB = action_management.a
+LIB = libactionmanagement.a
```
Why was this change made? Just curious because I think it might break a lot of people's make files who are linking your library

MTUC08 replied:
This was b/c our tests weren't compiling and I thought static libraries had to be of the form libactionmanagement.a since I was using -l. However, I later figured out that I should've been using -l: and reverted to the original name action_management.a

ecoble commented on src/action_management/src/actionmanagement.c:
```diff
@@ -50,58 +50,102 @@ void action_type_free(action_type_t *a)
     if (a->kind != ITEM) {
-        fprintf(stderr, "The action type provided is not of the correct kind.\n");
-        return FAILURE;
+        strcat(ret_string, "The action type provided is not of the correct kind");
```
This should theoretically never happen since CLI decides which of your functions to use based on the kind of the action, but doesn't hurt to double check I guess

MTUC08 replied:
Yeah, borja said that in class too, but its just a double check just in case.

ecoble commented on src/action_management/src/actionmanagement.c:
```diff
@@ -50,58 +50,102 @@ void action_type_free(action_type_t *a)
-int do_item_action(game_t *g, action_type_t *a, item_t *i)
+/* KIND 1
+ * See actionmanagement.h */
+char *do_item_action(game_t *g, action_type_t *a, item_t *i)
```
Perhaps not a change for this PR but have you considered using the new chiventure context struct that has the game in it? Might be good future proofing since I think eventually everyone will be passing around a context struct not a game struct

MTUC08 replied:
We'll definitely look into that for sprint 4

ecoble commented on src/action_management/src/get_actions.c:
```diff
@@ -31,15 +31,33 @@ static int NUM_ACTIONS = sizeof(valid_actions) / sizeof(action_type_t);
+/* THIS FUNCTION IS USED IN TEST_GET_ACTIONS.C
+ * BUT IS LEFT HERE FOR REFERENCE
+ * head: SHOULD BE GET_SUPPORTED_ACTIONS()
+ * query: SHOULD BE A VALID LOWERCASE ACTION STRING C_NAME, e.g. "go", "pull", "open"
```
Is there a reason this is lowercase? It might be better if it is uppercase for consistency, on CLI's end we are all upper case on commands and I don't want casing to cause more integration problems for us

MTUC08 replied:
Sorry about that, I think that was left over from when we briefly switched over to lower-case. I'll change it to remove the lowercase requirement.

lucyyutingli commented on src/action_management/include/actionmanagement.h:
```diff
@@ -87,10 +89,10 @@ int do_item_action(game_t *g, action_type_t *a, item_t *i);
+char *do_path_action(game_t *g, action_type_t *a, path_t *p);
 
 
 /* A function that executes KIND 5 actions (ACTION <item> <item>)
```
We no longer have 5 kinds of actions, needs to be changed to 3 (I think)

lucyyutingli commented on src/action_management/include/actionmanagement.h:
```diff
@@ -102,10 +104,10 @@ int do_path_action(game_t *g, action_type_t *a, path_t *p);
+ * - A success string upon success
  */
-int do_item_item_action(game_t *g, action_type_t *a, item_t *direct, item_t *indirect);
+char *do_item_item_action(game_t *g, action_type_t *a, item_t *direct, item_t *indirect);
```
I thought we decided to change it back to returning an int based off Borja's recommendation from the meeting? Lemme know if I'm wrong.

MTUC08 replied:
Thats going to be done in sprint 4, this is for the sprint 3 stuff

lucyyutingli commented on src/action_management/src/actionmanagement.c:
```diff
@@ -50,58 +50,102 @@ void action_type_free(action_type_t *a)
-int do_item_action(game_t *g, action_type_t *a, item_t *i)
+/* KIND 1
+ * See actionmanagement.h */
+char *do_item_action(game_t *g, action_type_t *a, item_t *i)
```
Char -> int, lemme same as above.

MTUC08 replied:
issue #245

lucyyutingli commented on src/action_management/src/actionmanagement.c:
```diff
@@ -50,58 +50,102 @@ void action_type_free(action_type_t *a)
-int do_path_action(game_t *g, action_type_t *a, path_t *p)
+/* KIND 2
+ * See actionmanagement.h */
+char *do_path_action(game_t *g, action_type_t *a, path_t *p)
```
Same as before, char -> int

MTUC08 replied:
again, later issue

mpcsh commented on src/action_management/Makefile:
```diff
@@ -14,14 +14,18 @@ RM = rm -f
 
 $(LIB): $(OBJS)
-	ar -r -o $@ $(OBJS)
+	$(AR) rcs $@ $^ 
```
What is `rcs`?

MTUC08 replied:
Our program was continuously crashing, saying the -r flag was incorrectly used, so I found a solution from this site: https://stackoverflow.com/questions/29714300/what-does-the-rcs-option-in-ar-do. rcs basically means replace, create, sort.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -50,58 +50,102 @@ void action_type_free(action_type_t *a)
+    assert(g->curr_player); // needed for sprint 4
     assert(a);
     assert(i);
+    char *ret_string = malloc(100); // buffer
```
`100` is a magic number here - there should be a `#define`d constant at the top of this file for the buffer size. Be sure to change the other `malloc` calls in this file too

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -50,58 +50,102 @@ void action_type_free(action_type_t *a)
-int do_item_action(game_t *g, action_type_t *a, item_t *i)
+/* KIND 1
+ * See actionmanagement.h */
+char *do_item_action(game_t *g, action_type_t *a, item_t *i)
```
Just for clarification: what we discussed during team meeting about having this function return `int` and using a `char *` out-parameter will be included in a fast-follow to this PR, correct?

MTUC08 replied:
Yes, I created a separate issue #245 to handle that, and CLI is also working on that branch to fix whatever that change will break so those changes will go straight to master along with the break.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -50,58 +50,102 @@ void action_type_free(action_type_t *a)
+    // checks if the action can be used on the item
+    int allowed = allowed_action(i, a->c_name);
+    if (allowed != SUCCESS) {
+        strcat(ret_string, "Action ");
```
Better way to do this: `sprintf(ret_string, "Action %s can't be requested on item %s", a->c_name, i->item_id)`. Read `man sprintf` for more info. Let's go ahead and change all the groups of `strcat`s in this file to `sprintf`

mpcsh commented on src/action_management/src/get_actions.c:
```diff
@@ -31,15 +31,33 @@ static int NUM_ACTIONS = sizeof(valid_actions) / sizeof(action_type_t);
     return temp;
 }
+
+/* THIS FUNCTION IS USED IN TEST_GET_ACTIONS.C
```
Commented code should be deleted

mpcsh commented on src/action_management/tests/test_get_actions.c:
```diff
@@ -0,0 +1,81 @@
+                 "search_supported_actions returned a pointer for invalid query \"fight\".\n");
+}
+
+Test(get_actions, count)
```
This test should be put first

MTUC08 commented:
Closing Statement:
In this PR we added the refactorization of our functions to the basic sprint 3 requirements (i.e. return a string to CLI) and added tests for get_supported actions.

Looking Forward:
We will follow this PR with a PR including the rest of our tests as well as the refactorization of our code to return ints for ease of error checking #289 . In Sprint 4, we will expand upon these requirements to include condition checking and effect application #188 #190 #191 and update the test accordingly #189 #192 #193 .

mpcsh commented:
Issue Score: ✓+

Comments:
Good work, but the closing statement only came well after the fact.
<!-- {"score": "check-plus"} -->
